modules/core/src/main/java/org/apache/ignite/internal/commandline/CommandHandler.java

```diff
@@ -1398,6 +1413,257 @@ else if (arg.getOperation() == VisorTxOperation.KILL)
+
+            if (verboseInfo != null) {
+                log(indent + "Mapping [type=" + verboseInfo.txMappingType() + "]:");
+
+                printTransactionMapping(indent + DOUBLE_INDENT, info, verboseInfo);
+            }
+            else {
+                log(indent + "Mapping [type=HISTORICAL]:");
+
+                log(indent + DOUBLE_INDENT + "State: " + info.getState());
+            }
+        }
+    }
+
+    /**
+     * Prints specific transaction mapping to output.
+     *
+     * @param indent Indent.
+     * @param info Info.
+     * @param verboseInfo Verbose info.
+     */
+    private void printTransactionMapping(String indent, VisorTxInfo info, TxVerboseInfo verboseInfo) {
+        log(indent + "XID version (UUID): " + info.getXid());
+        log(indent + "State: " + info.getState());
+
+        if (verboseInfo.txMappingType() == TxMappingType.REMOTE) {
+            log(indent + "Primary node: " + verboseInfo.dhtNodeId());
+            log(indent + "Primary node (consistent ID): " + verboseInfo.dhtNodeConsistentId());
+        }
+
+        if (!F.isEmpty(verboseInfo.localTxKeys())) {
+            log(indent + "Mapped keys:");
+
+            printTransactionKeys(indent + DOUBLE_INDENT, verboseInfo);
+        }
+    }
+
+    /**
+     * Prints keys of specific transaction mapping to output.
+     *
+     * @param indent Indent.
+     * @param verboseInfo Verbose info.
+     */
+    private void printTransactionKeys(String indent, TxVerboseInfo verboseInfo) {
+        for (TxVerboseKey txVerboseKey : verboseInfo.localTxKeys()) {
+            log(indent + (txVerboseKey.read() ? "Read" : "Write") +
+                " [lock=" + txVerboseKey.lockType() + "]: " + txVerboseKey.txKey());
+
+            if (txVerboseKey.lockType() == TxKeyLockType.AWAITS_LOCK)
+                log(indent + DOUBLE_INDENT + "Lock owner XID: " + txVerboseKey.ownerVersion());
+        }
+    }
+
+    /**
+     * Prints results of --tx --info to output in case requested transaction is not active.
+     *
+     * @param res Response.
+     */
+    private void printTxInfoHistoricalResult(Map<ClusterNode, VisorTxTaskResult> res) {
+        if (F.isEmpty(res))
+            log("Transaction was not found in history across the cluster.");
+        else {
+            log("Transaction was found in completed versions history of the following nodes:");
+
+            for (Map.Entry<ClusterNode, VisorTxTaskResult> entry : res.entrySet()) {
+                log(DOUBLE_INDENT + nodeDescription(entry.getKey()) + ':');
+                log(DOUBLE_INDENT + DOUBLE_INDENT + "State: " + entry.getValue().getInfos().get(0).getState());
+            }
+        }
+    }
+
+    /**
+     * Checks that all connectable cluster nodes have specific (or greater) product version.
+     *
+     * @param client Client.
+     * @param ver Version.
+     */
+    private static void validateProductVersion(GridClient client, IgniteProductVersion ver) throws GridClientException {
```

Comment:
Could be replaced by IgniteFeatures?
And I didn't get, would we test client versions?

Reply:
We can, but I don't think that IgniteFeatures usage is necessary here.
Regarding client versions: good catch, thank you! I've fixed the code, now it will check clients as well.